helix-tui/src/text.rs

```diff
@@ -229,6 +229,70 @@ impl<'a> Spans<'a> {
     pub fn width(&self) -> usize {
         self.0.iter().map(Span::width).sum()
     }
+
+    /// Constructs a styled `Span` from the given `string` and `style`
+    /// and appends it the to the `Spans`.
+    ///
+    /// ## Examples
+    ///
+    /// ```rust
+    /// # use helix_tui::text::{Span, Spans};
+    /// # use helix_view::graphics::{Color, Style};
+    /// let mut spans = Spans::new();
+    /// spans.push("Hello!", Style::default());
+    /// assert_eq!("Hello!", spans.0[0].content);
+    /// ```
+    pub fn push<T>(&mut self, string: T, style: Style)
+    where
+        T: Into<Cow<'a, str>>,
+    {
+        self.0.push(Span::styled(string, style));
+    }
+
+    /// Creates a new, empty `Spans`.
+    pub fn new() -> Spans<'a> {
+        Spans(vec![])
+    }
+
+    /// Constructs a single, styled `Span` with the given `string` and `style`
+    /// and wraps it in a `Spans`.
+    pub fn single<T>(string: T, style: Style) -> Spans<'a>
+    where
+        T: Into<Cow<'a, str>>,
+    {
+        Spans(vec![Span::styled(string, style)])
+    }
+
+    /// Intersperses `sep` between items of the given iterator then returns the concatenated Spans.
+    ///
+    /// ## Examples
+    ///
+    /// ```rust
+    /// # use helix_tui::text::{Span, Spans};
+    /// # use helix_view::graphics::Style;
+    /// let spans = Spans::join(
+    ///   vec![
+    ///     Spans::single("Hello", Style::default()),
+    ///     Spans::single("World!", Style::default()),
+    ///   ].into_iter(),
+    ///   &Spans::single(" ", Style::default()),
+    /// );
+    /// assert_eq!("Hello World!", String::from(spans));
+    /// ```
+    pub fn join<S>(mut spanses: S, sep: &Spans<'a>) -> Spans<'a>
+    where
+        S: Iterator<Item = Spans<'a>>,
+    {
+        let mut spans: Vec<Span> = Vec::new();
+        if let Some(head) = spanses.next() {
+            spans.extend(head.0);
+        }
+        spanses.for_each(|s| {
+            spans.extend(sep.0.clone());
+            spans.extend(s.0);
+        });
```

Comment:
It's a style nit but `for_each` is generally considered less idiomatic in Rust than regular `for` or `while` loops

```rust
while let Some(spans) = span_iter.next() {
    if !new_spans.is_empty() {
        new_spans.push(sep.clone());
    }
    new_spans.extend(spans);
}
```

(see the note in the rust docs: https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.for_each)

Reply:
Will change, thanks!

Reply:
Fixed in 8e8d860b